ethcore/src/engines/authority_round.rs

```diff
@@ -0,0 +1,411 @@
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! A blockchain engine that supports a non-instant BFT proof-of-authority.
+
+use std::sync::atomic::{AtomicUsize, AtomicBool, Ordering as AtomicOrdering};
+use std::sync::Weak;
+use std::time::{UNIX_EPOCH, Duration};
+use util::*;
+use ethkey::{verify_address, Signature};
+use rlp::{UntrustedRlp, View, encode};
+use account_provider::AccountProvider;
+use block::*;
+use spec::CommonParams;
+use engines::Engine;
+use header::Header;
+use error::{Error, BlockError};
+use evm::Schedule;
+use ethjson;
+use io::{IoContext, IoHandler, TimerToken, IoService, IoChannel};
+use service::ClientIoMessage;
+use transaction::SignedTransaction;
+use env_info::EnvInfo;
+use builtin::Builtin;
+
+/// `AuthorityRound` params.
+#[derive(Debug, PartialEq)]
+pub struct AuthorityRoundParams {
+	/// Gas limit divisor.
+	pub gas_limit_bound_divisor: U256,
+	/// Time to wait before next block or authority switching.
+	pub step_duration: Duration,
+	/// Valid authorities.
+	pub authorities: Vec<Address>,
+	/// Number of authorities.
+	pub authority_n: usize,
+}
+
+impl From<ethjson::spec::AuthorityRoundParams> for AuthorityRoundParams {
+	fn from(p: ethjson::spec::AuthorityRoundParams) -> Self {
+		AuthorityRoundParams {
+			gas_limit_bound_divisor: p.gas_limit_bound_divisor.into(),
+			step_duration: Duration::from_secs(p.step_duration.into()),
+			authority_n: p.authorities.len(),
+			authorities: p.authorities.into_iter().map(Into::into).collect::<Vec<_>>(),
+		}
+	}
+}
+
+/// Engine using `AuthorityRound` proof-of-work consensus algorithm, suitable for Ethereum
+/// mainnet chains in the Olympic, Frontier and Homestead eras.
+pub struct AuthorityRound {
+	params: CommonParams,
+	our_params: AuthorityRoundParams,
+	builtins: BTreeMap<Address, Builtin>,
+	transistion_service: IoService<BlockArrived>,
+	message_channel: Mutex<Option<IoChannel<ClientIoMessage>>>,
+	step: AtomicUsize,
+	proposed: AtomicBool,
+}
+
+impl Header {
+    fn step(&self) -> Result<usize, ::rlp::DecoderError> {
+        UntrustedRlp::new(&self.seal()[0]).as_val()
+    }
+
+    fn signature(&self) -> Result<Signature, ::rlp::DecoderError> {
+        UntrustedRlp::new(&self.seal()[1]).as_val::<H520>().map(Into::into)
+    }
+}
+
+
+trait AsMillis {
+	fn as_millis(&self) -> u64;
+}
+
+impl AsMillis for Duration {
```

Comment:
using the `time` crate will get you this for free as well as other time-related utilities/potentially more exact measurements.